modules/Channel.md

```diff
@@ -0,0 +1,112 @@
+	and	sends the `value` to the channel.
+
+	* `recv(out value : eltType) : bool` : This function waits until there is
+	something in the buffer and stores the received entity in `value`. It also
+	returns the status i.e., the operation is executed successfully or not.
+
+	* `close()` : This function is used to close the channel such that no more
+	values will be sent to it.
+
+* `iter these()` : An iterator to receive the values sent to the channel until
+it is closed.
+
+* `select` statements : This statement lets you wait for multiple channel
+operations and executes if anyone of them is ready for communication. `select`
+statements in Go are specifically designed to work with channels.
+Some key points are:
+
+	* `select` statements are commonly used in Go code when working with channels.
+	* The operations are atomic, i.e., checking a channel and performing `recv`
+		or `send` on it is considered as one operation.
+	* The cases are arranged in a random order to prevent starvation.
+
+For e.g.
+```
+// Select syntax in Go to wait on any one of the two channels for receiving
+// If every channel is blocked, the default case is executed.
+select {
+	case msg1 := <-chan1 :
+		fmt.Println("Received", msg1)
+	case msg2 := <-chan2 :
+		fmt.Println("Received", msg2)
+	default :
+		fmt.Println("None")
+}
+```
+
+```
+// Select statements can be used to implement try-send or try-recv
+// default case is executed when none of the cases are ready yet.
+select {
+	case val := <-chan1 :
+		fmt.Println("Received", val);
+	default :
+		fmt.Println("Channel not ready");
+}
+```
+
+### Some important questions related to the module-
+
+* Should this module be standard or package ?
+
+* What should be the name of the module ? Some names in other languages
+	* `chan` in Golang
+	* `crossbeam` in Rust
+
+* What should be the name of the channel data type? Should it be record or
+class ?
+	* Channel should always be passed as a reference therfore we could make
+	it a class or a record-wrapped class.
+
+* What should be the syntax of Channel `select` statements in Chapel ?
+Should we use Chapel's `select` statement for that? Some semantic differences
+between `select` in Chapel and Channels.
+
+	* Chapel's select is a conditional statement which is used to evaluate
+	cases against one statement specified after `select` keyword whereas
+	`select` in Channels work with multiple Channel operations.
+
+	* `select` for Channels can be blocking or non-blocking while `select`
+	in Chapel is always non-blocking.
+
+	* A Go-style `select` statement has no argument but a Chapel one does have an
+	argument. As a result these could be easily distinguished at the syntax level.
+
+* Should operators be used for `send` and `recv` routines ? The operators in Go
+ behave differently in case of receive operation
+	* `val := <-chan` will only return receive value.
+	* `val, ok := <-chan` will store received value in `val` and status in `ok`.
```

Comment:
I feel slightly nervous about introducing new operators and don't immediately find these notations obvious putting myself in the shoes of someone who's never seen this before (e.g., how does `:=` differ from `=`?  What does `<-` mean?).  For that reason, I might start by seeing what a method-based syntax would look like in part to understand what other options we have, and in part to ease people into understanding the semantics (since methods will have names and argument names that may make meaning clearer).  Then, as we get the design right in that approach, we can decide whether introducing operators would make the methods even clearer, and what those operators should look like / how they should relate to current Chapel operators.

Reply:
I agree starting with named functions sounds reasonable. If we need operators, I'd lean towards creating one-sided versions of the I/O operator `<~>` so it would be `val <~ chan` to `recv`.

To me, I think whether or not we need operators has to do with how the `select` statement is going to work. Will it be possible to implement clearly with named functions? I think we will learn more as we get further in designing and implementing it.

Reply:
I'd start with procs also. To me this doesn't (yet) seem "core" enough to justify syntactic changes. (I think I can envision implementing this entirely in terms of sync vars and procs, though the result would be clunky in some ways.)

Reply:
@gbtitus - we might need syntactic changes to have the select feature on the new channels work reasonably well. Previously I had thought we could implement it with sync variables but there are some tricky pieces about ordering and @Yudhishthira1406 convinced me that it was more complicated.